### PR TITLE
CLOUDP-238527: Streams watch secrets changes

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: config/crd/bases/atlas.mongodb.com_atlasdeployments.yaml [hunk 1]
```diff
@@ -366,6 +366,8 @@ spec:
                     description: A list of atlas search indexes configuration for
                       the current deployment
                     items:
+                      description: SearchIndex is the CRD to configure part of the
```
Comment (igor-karpukhin) [inline config/crd/bases/atlas.mongodb.com_atlasdeployments.yaml:369]: Looks like I forgot to include these changes into my PR :)

File: pkg/controller/atlasstream/atlasstream_instance_controller.go [hunk 2]
```diff
@@ -186,3 +192,51 @@ func (r *AtlasStreamsInstanceReconciler) findStreamInstancesForStreamConnection(
...
+			r.Log.Errorf("failed to list Atlas stream connections: %e", err)
+			return []reconcile.Request{}
+		}
+
+		if len(atlasStreamConnections.Items) == 0 {
+			return []reconcile.Request{}
+		}
+	}
+
+	requests := make([]reconcile.Request, 0, len(atlasStreamConnections.Items))
```
Comment (s-urbaniak) [inline pkg/controller/atlasstream/atlasstream_instance_controller.go:236]: nit: most likely `len(atlasStreamConnections.Items)` is not enough if multiple instances are referencing each connection but not a blocker.

Comment (s-urbaniak) [inline pkg/controller/atlasstream/atlasstream_instance_controller.go:236]:
```suggestion
	streamInstances := make([]reconcile.Request, 0, len(atlasStreamConnections.Items))
```

File: pkg/controller/atlasstream/atlasstream_instance_controller.go [hunk 3]
```diff
@@ -186,3 +192,51 @@ func (r *AtlasStreamsInstanceReconciler) findStreamInstancesForStreamConnection(
 
 	return requests
 }
+
+func (r *AtlasStreamsInstanceReconciler) findStreamInstancesForSecret(ctx context.Context, obj client.Object) []reconcile.Request {
```
Comment (s-urbaniak) [inline pkg/controller/atlasstream/atlasstream_instance_controller.go:196]:
```suggestion
func (r *AtlasStreamsInstanceReconciler) findStreamInstancesForCertificateSecret(ctx context.Context, obj client.Object) []reconcile.Request {
```

**Review by igor-karpukhin — APPROVED** [review]

File: pkg/controller/atlasstream/atlasstream_instance_controller.go [hunk 4]
```diff
@@ -186,3 +192,51 @@ func (r *AtlasStreamsInstanceReconciler) findStreamInstancesForStreamConnection(
...
+		),
+	}
+
+	err := r.Client.List(ctx, atlasStreamConnections, listOps)
+	if err != nil {
+		r.Log.Errorf("failed to list Atlas stream connections: %e", err)
+		return []reconcile.Request{}
+	}
+
+	if len(atlasStreamConnections.Items) == 0 {
```
Comment (s-urbaniak) [inline pkg/controller/atlasstream/atlasstream_instance_controller.go:217]: why is this mutually exclusive? if there are no credentials secrets referencing stream instances, there can still be certificate secrets pointing to connections. So if anything, this has to be additive.

Comment (helderjs) [reply pkg/controller/atlasstream/atlasstream_instance_controller.go]: Clarified externally

**Review by s-urbaniak — APPROVED** [review]

**Review by josvazg — APPROVED** [review]

**Pull request merged**